scipy/signal/spectral.py

```diff
@@ -267,13 +267,184 @@ def welch(x, fs=1.0, window='hanning', nperseg=256, noverlap=None, nfft=None,
+        Length of each segment.  Defaults to 256.
+    noverlap: int, optional
+        Number of points to overlap between segments. If None,
+        ``noverlap = nperseg / 2``.  Defaults to None.
+    nfft : int, optional
+        Length of the FFT used, if a zero padded FFT is desired.  If None,
+        the FFT length is `nperseg`. Defaults to None.
+    detrend : str or function or False, optional
+        Specifies how to detrend each segment. If `detrend` is a string,
+        it is passed as the ``type`` argument to `detrend`.  If it is a
+        function, it takes a segment and returns a detrended segment.
+        If `detrend` is False, no detrending is done.  Defaults to 'constant'.
+    return_onesided : bool, optional
+        If True, return a one-sided spectrum for real data. If False return
+        a two-sided spectrum. Note that for complex data, a two-sided
+        spectrum is always returned.
+    scaling : { 'density', 'spectrum' }, optional
+        Selects between computing the power spectral density ('density')
+        where Pxy has units of V**2/Hz if x and y are measured in V and 
+        computing the power spectrum ('spectrum') where Pxy has units of V**2
+        if x and y are measured in V. Defaults to 'density'.
+    axis : int, optional
+        Axis along which the CSD is computed for both inputs; the default is 
+        over the last axis (i.e. ``axis=-1``).
+
+    Returns
+    -------
+    f : ndarray
+        Array of sample frequencies.
+    Pxy : ndarray
+        Cross spectral density or cross power spectrum of x,y.
+
+    See Also
+    --------
+    periodogram: Simple, optionally modified periodogram
+    lombscargle: Lomb-Scargle periodogram for unevenly sampled data
+    welch: Power spectral density by Welch's method. [Equivalent to csd(x,x)]
+    coherence: Magnitude squared coherence by Welch's method. 
+
+    Notes
+    --------
+    By convention, Pxy is computed with the conjugate FFT of X multiplied by 
+    the FFT of Y. 
+
+    If the input series differ in length, the shorter series will be 
+    zero-padded to match.
+
+    An appropriate amount of overlap will depend on the choice of window
+    and on your requirements.  For the default 'hanning' window an
+    overlap of 50\% is a reasonable trade off between accurately estimating
+    the signal power, while not over counting any of the data.  Narrower
+    windows may require a larger overlap.
+
+    If `noverlap` is 0, this method is equivalent to Bartlett's method [2]_.
+
+    References
+    ----------
+    .. [1] P. Welch, "The use of the fast Fourier transform for the
+           estimation of power spectra: A method based on time averaging
+           over short, modified periodograms", IEEE Trans. Audio
+           Electroacoust. vol. 15, pp. 70-73, 1967.
+    .. [2] M.S. Bartlett, "Periodogram Analysis and Continuous Spectra",
+           Biometrika, vol. 37, pp. 1-16, 1950.
+    """
+
+    if y is x: 
```

Comment:
I feel that csd(x, x) should indeed return a purely real result every time, since the underlying math is the product of the FFT of x and its conjugate, so a user shouldn't expect a complex vector back, since they're essentially asking for a PSD. 

Note that this will only happen if x and y are the same object in memory; if they're different objects, but the same data (as in y=np.copy(x)), a complex array is returned in which all of the imaginary parts happen to be zero. 

To me, it would seem a little odd to allow a CSD method to be called with only one input argument, but it wouldn't be too tough to rejigger the code to allow a y=None input.